packages/yew/src/dom_bundle/listeners.rs

```diff
@@ -0,0 +1,696 @@
+                    _ => unreachable!(),
+                };
+                test_log!("unregistering listeners for {}", id);
+                Registry::with(|reg| reg.unregister(id));
+                *bundle = NoReg;
+            }
+            (None, NoReg) => {
+                test_log!("{}", &"unchanged empty listeners");
+            }
+        };
+    }
+}
+
+impl ListenerRegistration {
+    /// Register listeners and return their handle ID
+    fn register(el: &Element, pending: &[Option<Rc<dyn Listener>>]) -> Self {
+        Self::Registered(Registry::with(|reg| {
+            let id = reg.set_listener_id(el);
+            reg.register(id, pending);
+            id
+        }))
+    }
+
+    /// Remove any registered event listeners from the global registry
+    pub(super) fn unregister(&self) {
+        if let Self::Registered(id) = self {
+            Registry::with(|r| r.unregister(id));
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, Eq, PartialEq, Debug)]
+struct EventDescriptor {
+    kind: ListenerKind,
+    passive: bool,
+}
+
+impl From<&dyn Listener> for EventDescriptor {
+    fn from(l: &dyn Listener) -> Self {
+        Self {
+            kind: l.kind(),
+            passive: l.passive(),
+        }
+    }
+}
+
+/// Ensures global event handler registration.
+//
+// Separate struct to DRY, while avoiding partial struct mutability.
+#[derive(Default, Debug)]
+struct GlobalHandlers {
+    /// Events with registered handlers that are possibly passive
+    handling: HashSet<EventDescriptor>,
+
+    /// Keep track of all listeners to drop them on registry drop.
+    /// The registry is never dropped in production.
+    #[cfg(test)]
+    #[allow(clippy::type_complexity)]
+    registered: Vec<(ListenerKind, Closure<dyn Fn(web_sys::Event)>)>,
+}
+
+impl GlobalHandlers {
+    /// Ensure a descriptor has a global event handler assigned
+    fn ensure_handled(&mut self, desc: EventDescriptor) {
+        if !self.handling.contains(&desc) {
+            let cl = BODY.with(|body| {
+                let cl = Closure::wrap(
+                    Box::new(move |e: Event| Registry::handle(desc, e)) as Box<dyn Fn(Event)>
+                );
+                AsRef::<web_sys::EventTarget>::as_ref(body)
+                    .add_event_listener_with_callback_and_add_event_listener_options(
+                        &desc.kind.as_ref()[2..],
+                        cl.as_ref().unchecked_ref(),
+                        &{
+                            let mut opts = web_sys::AddEventListenerOptions::new();
+                            opts.capture(true);
+                            // We need to explicitly set passive to override any browser defaults
+                            opts.passive(desc.passive);
+                            opts
+                        },
+                    )
+                    .map_err(|e| format!("could not register global listener: {:?}", e))
+                    .unwrap();
+                cl
```

Comment:
As with the change to `attributes`, this code has just been moved. I'd have to dig a bit to find out why it does it like that.  This was introduced in #1542 . I can change it here if you want, otherwise would put it off to another PR.

Reply:
Since you're touching it anyway, I don't see why not change. It would make the code less error-prone

Reply:
`gloo::events` wants a `|e: &Event|` closure, currently it's a `|e: Event|` closure. Both `unchecked_ref` cast it, do you happen to know which one is correct?

Reply:
Not sure. Does it even matter? It's a JavaScript type and `web-sys` implements `Clone` on it using `wasm-bindgen`

Reply:
The `unchecked_ref` might coerce an  `Event` into a `&Event` if the cast is mistakenly used. I'm not 100% sure if it just happens to work "accidentally" because `Event` internally is just a pointer to the real thing and `wasm_bindgen` operations see the same thing either way.